File: Sources/SubGroupOperationQueue.m
//
//  SubGroupOperationQueue.m
//  APNSubGroupOperationQueue
//
//  Created by André Pacheco Neves on 03/04/16.
//  Copyright © 2016 André Pacheco Neves. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

#import "SubGroupOperationQueue.h"

#pragma mark -
#pragma mark APNCompletionOperation

@interface APNCompletionOperation : NSBlockOperation
@end

@implementation APNCompletionOperation
@end

#pragma mark -
#pragma mark APNSubGroupOperationQueue

@interface APNSubGroupOperationQueue ()

@property(nonatomic, strong) dispatch_queue_t queue;
@property(nonatomic, strong) NSMutableDictionary<id<NSCopying>, NSArray<__kindof NSOperation *> *> *subGroups;

@end

@implementation APNSubGroupOperationQueue

- (instancetype)init {
    self = [super init];
    if (self) {
        self.queue = dispatch_queue_create("com.p4checo.APNSubGroupOperationQueue.queue", DISPATCH_QUEUE_CONCURRENT);
        self.subGroups = [NSMutableDictionary new];
    }
    return self;
}

#pragma mark -
#pragma mark Public

#pragma mark -
#pragma mark SubGroup operation scheduling

- (void)addOperation:(NSOperation *)op withKey:(id<NSCopying>)key {
    NSParameterAssert(op);
    NSParameterAssert(key);

    __block NSArray<NSOperation *> *opPair;

    dispatch_barrier_sync(self.queue, ^{
        NSArray<NSOperation *> *subGroup = self.subGroups[key] ?: @[];

        APNCompletionOperation *competionOp = [self addDependenciesOnOperation:op withKey:key inSubGroup:subGroup];

        opPair = @[ op, competionOp ];

        self.subGroups[key] = [subGroup arrayByAddingObjectsFromArray:opPair];
    });

    [self addOperations:opPair waitUntilFinished:false];
}

- (void)addOperations:(NSArray<__kindof NSOperation *> *)ops withKey:(id<NSCopying>)key waitUntilFinished:(BOOL)wait {
    NSParameterAssert(ops.count);
    NSParameterAssert(key);

    __block NSMutableArray<NSOperation *> *newOps = [NSMutableArray array];

    dispatch_barrier_sync(self.queue, ^{
        NSMutableArray<NSOperation *> *subGroup = [self.subGroups[key] ?: @[] mutableCopy];

        for (NSOperation *op in ops) {
            APNCompletionOperation *competionOp = [self addDependenciesOnOperation:op withKey:key inSubGroup:subGroup];

            NSArray<NSOperation *> *opPair = @[ op, competionOp ];
            [newOps addObjectsFromArray:opPair];
            [subGroup addObjectsFromArray:opPair];
        }

        self.subGroups[key] = [subGroup copy];
    });

    [self addOperations:newOps waitUntilFinished:wait];
}

- (void)addOperationWithBlock:(void (^)(void))block andKey:(id<NSCopying>)key {
    NSParameterAssert(block);
    NSParameterAssert(key);

    [self addOperation:[NSBlockOperation blockOperationWithBlock:block] withKey:key];
}

#pragma mark -
#pragma mark SubGroup querying

- (NSArray<__kindof NSOperation *> *)subGroupOperationsForKey:(id<NSCopying>)key {
    NSParameterAssert(key);

    static NSPredicate *classPredicate;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        classPredicate = [NSPredicate predicateWithFormat:@"class != %@", [APNCompletionOperation class]];
    });

    __block NSArray<NSOperation *> *ops;

    dispatch_sync(self.queue, ^{
        ops = [self.subGroups[key] filteredArrayUsingPredicate:classPredicate];
    });

    return ops ?: @[];
}

#pragma mark -
#pragma mark Setters

- (void)setMaxConcurrentOperationCount:(NSInteger)maxConcurrentOperationCount {
    NSAssert(maxConcurrentOperationCount != 1,
             @"`APNSubGroupOperationQueue` must be concurrent to provide any benefit over a serial queue! 🙃");

    super.maxConcurrentOperationCount = maxConcurrentOperationCount;
}

#pragma mark -
#pragma mark Private

NS_ASSUME_NONNULL_BEGIN

- (APNCompletionOperation *)addDependenciesOnOperation:(NSOperation *)op
                                               withKey:(id<NSCopying>)key
                                            inSubGroup:(NSArray<__kindof NSOperation *> *)subGroup {
    NSParameterAssert(op);
    NSParameterAssert(key);
    NSParameterAssert(subGroup);

    APNCompletionOperation *completionOp = [self completionOperationForOperation:op withKey:key];
    [completionOp addDependency:op];

    // new operations only need to depend on the group's last operation
    NSOperation *lastOp = subGroup.lastObject;
    lastOp ? [op addDependency:lastOp] : nil;

    return completionOp;
}

- (APNCompletionOperation *)completionOperationForOperation:(NSOperation *)op withKey:(id<NSCopying>)key {
    NSParameterAssert(op);
    NSParameterAssert(key);

    APNCompletionOperation *completionOp = [APNCompletionOperation new];

    __weak typeof(completionOp) weakCompletionOp = completionOp;

    [completionOp addExecutionBlock:^{
        dispatch_barrier_sync(self.queue, ^(void) {
            __strong typeof(weakCompletionOp) strongCompletionOp = weakCompletionOp;

            if (!strongCompletionOp) {
                NSAssert(NO, @"💥: completion operation must not be nil");
                return;
            }

            NSArray<NSOperation *> *subGroup = self.subGroups[key];

            NSAssert(subGroup, @"💥: A subgroup must exist in the dicionary for the finished operation's key!");
            NSAssert([op isEqual:subGroup[0]],
                     @"💥: Finished operation must be the first element in the subgroup's array");
            NSAssert([strongCompletionOp isEqual:subGroup[1]],
                     @"💥: completionOp must be the second element in the subgroup's array");

            self.subGroups[key] = subGroup.count == 2 ? nil : ({
                NSMutableArray<NSOperation *> *newSubGroup = [subGroup mutableCopy];
                [newSubGroup removeObjectsInRange:NSMakeRange(0, 2)];
                [newSubGroup copy];
            });
        });
    }];

    return completionOp;
}

NS_ASSUME_NONNULL_END

@end

File: Sources/SubGroupOperationQueue.swift
//
//  SubGroupOperationQueue.swift
//  APNSubGroupOperationQueue
//
//  Created by André Pacheco Neves on 27/03/16.
//  Copyright © 2016 André Pacheco Neves. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

import Foundation

/** 
 `SubGroupOperationQueue` is an `OperationQueue` subclass which allows scheduling operations in serial subgroups inside
 a concurrent queue. 
 
 The subgroups are stored as a `[Key : [Operation]]`, and each subgroup array contains all the scheduled subgroup's
 operations which are pending and executing. Finished `Operation`s are automatically removed from the subgroup after 
 completion.
*/
public class SubGroupOperationQueue<Key: Hashable>: OperationQueue {

    fileprivate let subGroups = OperationSubGroupMap<AnyHashable>()
    
    /** 
     The maximum number of queued operations that can execute at the same time.
     
     - warning: This value should be `!= 1` (serial queue), otherwise this class provides no benefit.
    */
    override public var maxConcurrentOperationCount: Int {
        get { return super.maxConcurrentOperationCount }
        set {
            assert(newValue != 1, "`SubGroupQueue` must be concurrent to provide any benefit over a serial queue! 🙃")
            super.maxConcurrentOperationCount = newValue
        }
    }

    // MARK: - Public

    /**
     Adds the specified operation to the queue, and registers it the subgroup identified by `key`.
     
     Once added, the operation will only be executed after all currently existing operations in the same subgroup finish
     * executing (serial processing), but can be executed concurrently with other subgroup's operations.
     
     - parameter op:  The `Operation` to be added to the queue.
     - parameter key: The subgroup's identifier key.
     */
    public func addOperation(_ op: Operation, withKey key: Key) {
        addOperations(subGroups.register(op, withKey: key), waitUntilFinished: false)
    }
    
    /**
     Adds the specified operations to the queue, and registers them the subgroup identified by `key`. 
     The order in which the operations are processed is the same as the array's.
     
     Once added, the operations will be executed in order after all currently existing operations in the same subgroup
     finish executing (serial processing), but can be executed concurrently with other subgroup's operations.
     
     - parameter ops:  The `[Operation]` to be added to the queue
     - parameter key:  The subgroup's identifier key
     - parameter wait: If `true`, the current thread is blocked until all of the specified operations finish executing. 
     If `false`, the operations are added to the queue and control returns immediately to the caller.
     */
    public func addOperations(_ ops: [Operation], withKey key: Key, waitUntilFinished wait: Bool) {
        addOperations(subGroups.register(ops, withKey: key), waitUntilFinished: wait)
    }
    
    /**
     Wraps the specified block in an operation object, adds it to the queue and and registers it the subgroup identified
     by `key`.
     
     Once added, the operation will only be executed after all currently existing operations in the same subgroup finish
     executing (serial processing), but can be executed concurrently with other subgroup's operations.
     
     - parameter block: The block to execute from the operation.
     - parameter key:   The subgroup's identifier key.
     */
    public func addOperation(_ block: @escaping () -> Void, withKey key: Key) {
        addOperations(subGroups.register(block, withKey: key), waitUntilFinished: false)
    }
    
    // MARK: SubGroup querying
    
    /**
     Return a snapshot of currently scheduled (i.e. non-finished) operations of the subgroup identified by `key`.
     
     - parameter key: The subgroup's identifier key.
     
     - returns: An `[Operation]` containing a snapshot of all currently scheduled (non-finished) subgroup operations.
     */
    public subscript(key: Key) -> [Operation] {
        return subGroups[key]
    }
    
    /**
     Return a snapshot of currently scheduled (i.e. non-finished) operations of the subgroup identified by `key`.
     
     - parameter key: The subgroup's identifier key.
     
     - returns: An `[Operation]` containing a snapshot of all currently scheduled (non-finished) subgroup operations.
     */
    public func subGroupOperations(forKey key: Key) -> [Operation] {
        return subGroups[key]
    }
}
